dataflow; expr: Reuse Row allocations for SafeMfp evaluations

antiguru opened the pull request:
Currently, evaluating a SafeMfp either allocates a row or returns an
iterator. With this change, a third approach reusing an allocated Row is
added.

Signed-off-by: Moritz Hoffmann <mh@materialize.com>

### Motivation

This PR refactors existing code to avoid computing the length of a row repeatedly and instead reuse allocations.

### Description

The current `SafeMfpPlan::evaluate` function calculates the required length for each row individually. In many cases, this length remains constant throughput the lifetime of a dataflow. With this change, we provide a row with an allocation to build evaluate the MFP into. The final row is then copied, saving the cost to compute the required size.

### Checklist

- [ ] This PR has adequate test coverage.
- [ ] This PR adds a release note for any user-facing behavior changes.

frankmcsherry commented on src/dataflow/src/render/join/linear_join.rs:
```diff
@@ -299,6 +301,10 @@ where
         closure: JoinClosure,
         errors: &mut Vec<Collection<G, DataflowError>>,
     ) -> Collection<G, Row> {
+        println!(
+            "differential join stream/lookup key: {:?}\t{:?}",
+            stream_key, lookup_key
+        );
```
Probably to delete?

antiguru replied:
Argh, that was my attempt at making progress for #6672 :-D. I'll get it out of here.

frankmcsherry commented on src/expr/src/linear.rs:
```diff
@@ -1161,6 +1161,27 @@ pub mod plan {
         /// As the evaluation exits early with failed predicates, it may
         /// miss some errors that would occur later in evaluation.
         #[inline(always)]
+        pub fn evaluate_into<'a>(
```
I'd add some doc comments here about `row` and what happens. E.g., in this case "not cleared first, but emptied if the result is `Ok(Some(_))`". I'd be tempted to tweak the logic to try and make that more consistent (perhaps empty it first, so that we never end up with half-filled data?), or very explicitly document it as containing "undefined-ish" contents after the call.

antiguru replied:
I added a comment to describe this behavior, and added a call to `Row::clean()` before extending.

frankmcsherry reviewed (approved):
I think this looks good, and you could probably straight up delete `evaluate` (the replacement is I think almost a strict improvement, but a bit noisier for some uses). The only nit I have is that the contract about the contents of `row` is not clear, which is important because the code is called in a loop in various places. In the case of an `Err` or a `Some(None)`, will it be half-populated? Probably not I think, but .. best to be clear.

antiguru commented:
Thanks for the review! I removed `evaluate` and updated the documentation to be clear about what happens to the state of row, and rows are cleared to defend against possible leftovers. This we could revisit if there is a good case for extending partially filled rows.
